src/scheduler/Executor.cpp

```diff
@@ -497,14 +501,24 @@ void Executor::threadPoolThread(std::stop_token st, int threadPoolIdx)
         // Set the return value
         msg.set_returnvalue(returnValue);
 
-        // Decrement the task count. If the counter reaches zero at this point,
-        // all other tasks must also have reached this point, hence we can
-        // guarantee that the other tasks have also finished, and perform any
-        // merging or tidying up.
+        // Decrement the task count. If we are the last thread in a batch of
+        // we can either be in the main host, in which case there's still the
+        // main thread in the same Executor, or in a remote host, in which case
+        // we need to only release the Executor after we are done. If we
+        // are the main thread, we will always reset and release the executor
         std::atomic_thread_fence(std::memory_order_release);
-        int oldTaskCount = batchCounter.fetch_sub(1);
-        assert(oldTaskCount >= 0);
-        bool isLastInBatch = oldTaskCount == 1;
+        int oldTaskCount = 0;
+        bool isLastThreadInBatch = false;
+        bool isLastThreadInExecutor = false;
+        if (isThreads) {
+            oldTaskCount = threadBatchCounter.fetch_sub(1);
+            isLastThreadInBatch = oldTaskCount == 1;
+            isLastThreadInExecutor = batchCounter.load(std::memory_order_release) == 0;
+        } else {
+            oldTaskCount = batchCounter.fetch_sub(1);
+            isLastThreadInExecutor = oldTaskCount == 1;
+        }
+        assert(oldTaskCount >= 1);
```

Comment:
`taskCount` should be > 0, thus `oldTaskCount` should always be greater than `1`. This assertion would have caught the bug.